### ISSUE TITLE
Implementing oversampling in plug-ins.

### ISSUE THREAD

**Issue by Hussam-Eddin-Alhomsi** [LMMS/lmms#3605]
• Instrument plug-ins (synthesizers only):
When playing high notes on a synthesizer using a harmonically rich wave like a saw wave, there are probably some frequencies that will go above the nyquist frequency, so aliasing occurs.

Just play the note C8 on TripleOscillator using a saw wave to hear the aliasing noise.

My idea about the GUI implementation is a bunch of radio buttons in the plugin tab of each synthesizer offering oversampling factors in the power of two that's "1x" (no oversampling), "2x", "4x", "8x", "16x", "32x" and "64x".

• Effect plug-ins: 
There are 2 types of plug-ins that would benefit from using oversampling (as far as I know).

(1) Waveshapers, distorters, overdrivers, bitcrushers and Lo-fi effects (or anything that generates a lot of harmonics).
When using any of the effects above on a sound, extra harmonics are being added to it, and if these harmonics go above the nyquist frequency, aliasing occurs.

A spectrogram of a heavily waveshaped sine wave. (Note the aliasing noise highlighted in yellow before the fundamental frequency of the sine wave.)
![distorted sine wave](https://cloud.githubusercontent.com/assets/25503477/26704782/30472e22-473a-11e7-9854-3ef5930b41cd.png)
A spectrogram of the same waveshaped sine wave but the waveshaper has 16x oversampling active. (Note that most of the aliasing that was present before activating oversampling is gone.)
![distorted sine wave with 16x oversampling](https://cloud.githubusercontent.com/assets/25503477/26704785/34b8acd8-473a-11e7-839e-220765215709.png)

My idea about the GUI implementation is a slider in each of these plug-ins mentioned earlier offering oversampling factors like "1x" (no oversampling), "2x", "4x", "8x", "16x", "32x" and "64x".

(2) Equalizers.
When a band in an equalizer gets close to the nyquist frequency, its shape distorts.
The distortion can happen with any band type.

A normal low-pass filter at 1000 Hz looks like this:
![lpf 1000 hz](https://cloud.githubusercontent.com/assets/25503477/26700675/ea4781d2-4726-11e7-9181-0b1ea01809da.png)
The same low-pass filter with the same resonance but at 16000 Hz looks like this:
![lpf 16000 hz](https://cloud.githubusercontent.com/assets/25503477/26700730/23bff5f2-4727-11e7-97f0-4439165ad109.png)

The distortion isn't a bug, it's normal.
Using oversampling will prevent this distortion and improve audio quality, particularly at high frequencies (i.e, 4000 Hz and over).
My idea about the implementation is an off/on "Oversampling" button on each equalizer plug-in effect that uses a specific oversampling factor that is just enough to prevent the distortion without introducing excess CPU load.

NOTE: Using oversampling increases CPU load.

**Comment by PaulBatchelor** [comment]
Have you looked at the oversampling options in the export menu?

**Comment by Hussam-Eddin-Alhomsi** [comment]
Yes. But it doesn't even work at the moment, but even if it does work, it's on export only.
There should be an option to use it before exporting.

**Comment by grejppi** [comment]
IIRC the `Oscillator` class has band-limited waveforms available, which are available in LB302 and Monstro, but they are not enabled for TripleOscillator.

**Comment by Hussam-Eddin-Alhomsi** [comment]
Without effects post-processing, that would eliminate aliasing in instrument plug-ins that don't have have FM or AM/RM.
The problem is that some instruments have FM or AM/RM which can add overtones to the generated band limited waveform which can cause aliasing.
In an instrument plug-in like TripleOscillator that has both FM and AM, I think we should enable the band limited waveforms and an oversampling option.

**Comment by PaulBatchelor** [comment]
This feature would require a lot of internal refactoring of the plugin and classes, as well as introduce new UI. It's a very tall order, but the code is there if you'd like to take a stab at it and make a PR. I'm not also entirely convinced that this would be an effective new feature. Oversampling is very much a brute-force approach to alias reduction. I'd be more keen on approaching more specific solutions to alias reduction for particular plugins and effects, rather than try cram a "one shoe fits all" feature. Or better yet, just build new instruments and effects that sound better.

**Comment by Rossmaxx** [comment]
Oversampling is gone entirely since #7228

**Comment by Rossmaxx** [comment]
@sakertooth close?

**Comment by zonkmachine** [comment]
> Oversampling is gone entirely since #7228

The description of that PR says `should really be an option on a per-plugin basis,` which this issue is all about. If someone want's to implement oversampling in a plugin or two, why not?

**Comment by Rossmaxx** [comment]
Oops, another misread.
